xfrd.c

```diff
@@ -2383,78 +2525,255 @@ find_same_master_notify(xfrd_zone_type* zone, int acl_num_nfy)
+		if(committed && zone->latest_xfr) {
+			xfrd_schedule_zone_xfrs(
+				zone, zone->latest_xfr, committed);
+		}
+		return; /* continue */
+	}
+
+	/* zone transfer list is in descending order. since zone
+	   verification is done after all transfers have been applied,
+	   the state of the most recently tried transfer is the
+	   state for the rest of the list if it was "ok". the same is not
+	   true for "drop" as that may be caused by the most recent
+	   zone transfer (or last). "corrupt" can occur before a
+	   transfer reported as "ok" or "drop" as the transfer itself
+	   may have been "corrupt" */
+
+	/* if the state of the most recent transfer is "corrupt", any
+	   incremental transfer that follows must be discarded and the zone
+	   must be refreshed, unless of course a full transfer is already
+	   awaiting scheduling */
+	if(xfr->state == xfrd_xfr_corrupt ||
+	   xfr->state == xfrd_xfr_inconsistent)
 	{
-		/* zone has a disk soa, and no nsd soa or a different nsd soa */
-		if(zone->soa_disk_acquired != 0 &&
-			(zone->soa_nsd_acquired == 0 ||
-			zone->soa_disk.serial != zone->soa_nsd.serial))
-		{
-			if(zone->soa_disk_acquired <
-				xfrd->reload_cmd_last_sent)
-			{
-				/* this zone should have been loaded, since its disk
-				   soa time is before the time of the reload cmd. */
-				xfrd_soa_type dumped_soa = zone->soa_disk;
-				log_msg(LOG_ERR, "xfrd: zone %s: soa serial %u "
-						 		 "update failed, restarting "
-						 		 "transfer (notified zone)",
-					zone->apex_str, (unsigned)ntohl(zone->soa_disk.serial));
-				/* revert the soa; it has not been acquired properly */
-				if(zone->soa_disk_acquired == zone->soa_nsd_acquired) {
-					/* this was the same as served,
-					 * perform force_axfr , re-download
-					 * same serial from master */
-					zone->soa_disk_acquired = 0;
-					zone->soa_nsd_acquired = 0;
-				} else {
-					/* revert soa to the one in server */
-					zone->soa_disk_acquired = zone->soa_nsd_acquired;
-					zone->soa_disk = zone->soa_nsd;
-				}
-				/* pretend we are notified with disk soa.
-				   This will cause a refetch of the data, and reload. */
-				xfrd_handle_incoming_notify(zone, &dumped_soa);
-				xfrd_set_timer_refresh(zone);
-			} else if(zone->soa_disk_acquired >= xfrd->reload_cmd_last_sent) {
-				/* this zone still has to be loaded,
-				   make sure reload is set to be sent. */
-				if(xfrd->need_to_send_reload == 0 &&
-					xfrd->reload_added == 0) {
-					log_msg(LOG_ERR, "xfrd: zone %s: needs "
-									 "to be loaded. reload lost? "
-									 "try again", zone->apex_str);
-					xfrd_set_reload_timeout();
-				}
+		next_xfr = xfr->next;
+		while(next_xfr) {
+			/* transfer must be complete to determine if it is
+			   incremental */
+			if(next_xfr->acquired > 0 && next_xfr->msg_is_ixfr) {
+				xfrd_discard_zone_xfr(zone, next_xfr);
+				next_xfr = xfr->next;
+			} else {
+				break;
 			}
 		}
+		/* refresh if no full transfer was awaiting scheduling */
```

Comment:
At this point the comment says you look at if a full transfer was waiting, but the code did not check for that.  It removed ixfrs, but did not check if there is an AXFR or if the loop ended nonNULL.

Reply:
Correct. I had that in there at some point (comment probably remained), but (from memory) `msg_is_ixfr` might be certain, it didn't seem to be that if that's `false` it's necessarily an axfr. Just doing a refresh in all scenarios seemed like a better decision. Of course, I might have overlooked something or just doing the wrong check all together.

Reply:
It may be required to do a full AXFR after DNSSEC corrupt contents.  Otherwise the master may reply with a smaller incremental update that moves to the failed position every time, it would have to reply with a concatenation or compaction of all updates so that when the zone DNSSEC failure is solved we actually get it.  Probably IXFR is good enough but it depends on the master service of IXFR, right now we rely on the master sending an IXFR update to the latest version, and I think that is actually OK.